crates/noirc_evaluator/src/ssa_refactor/ir/instruction/call.rs

```diff
@@ -0,0 +1,340 @@
+
+        BlackBoxFunc::FixedBaseScalarMul | BlackBoxFunc::SchnorrVerify | BlackBoxFunc::Pedersen => {
+            // Currently unsolvable here as we rely on an implementation in the backend.
+            SimplifyResult::None
+        }
+        BlackBoxFunc::RANGE | BlackBoxFunc::RecursiveAggregation => SimplifyResult::None,
+    }
+}
+
+fn make_constant_array(dfg: &mut DataFlowGraph, results: Vec<FieldElement>, typ: Type) -> ValueId {
+    let result_constants = vecmap(results, |element| dfg.make_constant(element, typ.clone()));
+
+    let typ = Type::Array(Rc::new(vec![typ]), result_constants.len());
+    dfg.make_array(result_constants.into(), typ)
+}
+
+/// Returns a Value::Array of constants corresponding to the limbs of the radix decomposition.
+fn constant_to_radix(
+    endian: Endian,
+    field: FieldElement,
+    radix: u32,
+    limb_count: u32,
+    dfg: &mut DataFlowGraph,
+) -> ValueId {
+    let bit_size = u32::BITS - (radix - 1).leading_zeros();
+    let radix_big = BigUint::from(radix);
+    assert_eq!(BigUint::from(2u128).pow(bit_size), radix_big, "ICE: Radix must be a power of 2");
+    let big_integer = BigUint::from_bytes_be(&field.to_be_bytes());
+
+    // Decompose the integer into its radix digits in little endian form.
+    let decomposed_integer = big_integer.to_radix_le(radix);
+    let mut limbs = vecmap(0..limb_count, |i| match decomposed_integer.get(i as usize) {
+        Some(digit) => FieldElement::from_be_bytes_reduce(&[*digit]),
+        None => FieldElement::zero(),
+    });
+    if endian == Endian::Big {
+        limbs.reverse();
+    }
+
+    // For legacy reasons (see #617) the to_radix interface supports 256 bits even though
+    // FieldElement::max_num_bits() is only 254 bits. Any limbs beyond the specified count
+    // become zero padding.
+    let max_decomposable_bits: u32 = 256;
+    let limb_count_with_padding = max_decomposable_bits / bit_size;
+    while limbs.len() < limb_count_with_padding as usize {
+        limbs.push(FieldElement::zero());
+    }
+
+    make_constant_array(dfg, limbs, Type::unsigned(bit_size))
+}
+
+fn to_u8_vec(dfg: &DataFlowGraph, values: im::Vector<Id<Value>>) -> Vec<u8> {
+    values
+        .iter()
+        .map(|id| {
+            let field = dfg
+                .get_numeric_constant(*id)
+                .expect("value id from array should point at constant");
+            *field.to_be_bytes().last().unwrap()
+        })
+        .collect()
+}
+
+fn array_is_constant(dfg: &DataFlowGraph, values: &im::Vector<Id<Value>>) -> bool {
+    values.iter().all(|value| dfg.get_numeric_constant(*value).is_some())
+}
+
+fn simplify_hash(
+    dfg: &mut DataFlowGraph,
+    arguments: &[ValueId],
+    hash_function: fn(&[u8]) -> Result<[u8; 32], BlackBoxResolutionError>,
+) -> SimplifyResult {
+    match dfg.get_array_constant(arguments[0]) {
+        Some((input, _)) if array_is_constant(dfg, &input) => {
+            let input_bytes: Vec<u8> = to_u8_vec(dfg, input);
```

Comment:
In `simplify_hash` I'm having to check that all the value ids contained in a array constant are themselves constant. This seems a bit odd to me as I would have thought that the former implies the latter but if I remove the `array_is_constant` check then I get panics in `to_u8_vec`.

@jfecher does this look correct to you? I want to check that I'm not just masking an underlying issue with this check.

Reply:
This is intended, yes. A constant array implies that we know all the elements of an array so we can optimize away array get and set operations even if the elements themselves may be parameters or other non-constant values.